### ISSUE TITLE
Potentially upstream the improvements to PCL?

### ISSUE THREAD

**Issue by SteveMacenski** [koide3/ndt_omp#19]
Hi,

your readme mentions that it is 10x faster than PCL, any consideration of upstreaming it to PCL so that everyone can share that improvement? 

CC kunaltyagi

**Comment by kunaltyagi** [comment]
A PR would be welcome.

Please bear in mind a PR before GSoC begins or towards end of GSoC would be preferred. (Reasons being time and a GSoC project on new API to integrate SSE and non SSE code together)

**Comment by koide3** [comment]
Hi @SteveMacenski @kunaltyagi ,
I believe there is no major concerns about upstreaming this module to PCL.
Some minor concerns are:
- namespace and class names should be changed
- the code should be updated to follow the recent PCL registration interface convention

I would like to update the code and make a PR on the PCL repository.

**Comment by SteveMacenski** [comment]
Awesome! That sounds really great. Thanks for that help, it will go a _long_ way.

**Comment by SteveMacenski** [comment]
Any movement here? 

We're looking at implementing a NDT-MCL algorithm for Navigation2 to replace AMCL and we'd love to make use of this speed up!

**Comment by SteveMacenski** [comment]
(we're also open for any help / collaboration on that front)

**Comment by kunaltyagi** [comment]
Current status: One refactor PR is ongoing: https://github.com/PointCloudLibrary/pcl/pull/4180 (waiting on my review, perhaps by this weekend)

One has been merged, and NDT OMP PR hasn't been created yet (IIRC)

**Comment by koide3** [comment]
Yes, we have been working on refactoring of the original NDT implementation. After it approved, I'll open a PR for NDT OMP with all the corrections made at PointCloudLibrary/pcl#4180.

**Comment by SteveMacenski** [comment]
@koide3 what's the status of this? Looks like that PR is in so it should be ready?

**Comment by koide3** [comment]
@SteveMacenski 
I made a PR for the multi-threaded NDT, and probably it will be merged after a few review steps https://github.com/PointCloudLibrary/pcl/pull/4277
I'll inform you here once it's merged
